### PR TITLE
Restrict benchmarking to functions

### PR DESCRIPTION
**Pull request by mbauman**
Consider the function's arguments to be setup expressions, which are executed in an outer closure. The resulting arguments are then passed through a function barrier to the actual benchmarking function, which ensures that Julia has inferred concrete types for each of them.

Note that this removes the (at)noinline'd inner function.  This means that if the function being benchmarked is simple enough (or manually flagged) to inline, LLVM may eliminate the loop.  This reintroduces #5.

### CODE REVIEW

File: src/benchmarkable.jl [hunk 1]
```diff
@@ -27,33 +27,36 @@
...
-        end
-        $decls
         function $(esc(name))(
                 s::Samples,
                 n_samples::Integer,
                 evaluations::Integer,
             )
+            bench(s, n_samples, evaluations, $(map(esc, expr.args[2:end])...))
+        end
+        function bench(
```
Comment (johnmyleswhite) [inline src/benchmarkable.jl]: This doesn't need a gensym if you run many different benchmarks on different functions and different argument types in a row?

**Comment by mbauman** [comment]
I'm not sure whether we should remove the `@noinline`'d inner function.

Pros to removing it:
- Simpler
- Behaves exactly as though you called the function yourself inside a loop
- You can see how well LLVM can identify what's going on and if it's able to remove some work

Cons to removing it:
- LLVM can optimize it _too_ well and totally remove the entire benchmarking loop.
- Repeated calls isn't a typical use-case for an inlined function. The kinds of optimizations LLVM will make here will be very different than the kinds of optimizations that would normally occur.

File: src/benchmarkable.jl (hunk 1 above)
Comment (mbauman) [inline src/benchmarkable.jl]: Macro hygiene handles this for us:

``` jl
julia> macro f()
         :(f() = 1)
       end

julia> f1 = @f()
#15#f (generic function with 1 method)

julia> f2 = @f()
#16#f (generic function with 1 method)
```

**Comment by mbauman** [comment]
This does solve @staticfloat's use case in #22:

``` jl
julia> let
       x = 10
       @benchmark sin(x)
       end
================ Benchmark Results ========================
     Time per evaluation: 23.76 ns [23.54 ns, 23.98 ns]
Proportion of time in GC: 0.00% [0.00%, 0.00%]
        Memory allocated: 0.00 bytes
   Number of allocations: 0 allocations
       Number of samples: 9601
   Number of evaluations: 18827301
         R² of OLS model: 0.978
 Time spent benchmarking: 0.53 s


julia> @benchmark sin(x)
================ Benchmark Results ========================
     Time per evaluation: 23.85 ns [23.60 ns, 24.11 ns]
Proportion of time in GC: 0.00% [0.00%, 0.00%]
        Memory allocated: 0.00 bytes
   Number of allocations: 0 allocations
       Number of samples: 9601
   Number of evaluations: 18827301
         R² of OLS model: 0.971
 Time spent benchmarking: 0.53 s
```

**Comment by staticfloat** [comment]
I'm confused; shouldn't the second benchmark above fail? There should be no "x" in scope to pass in to sin().

**Comment by mbauman** [comment]
there was an `x=10` in global scope, too, I just missed copying it in.

**Comment by sglyon** [comment]
Could we somehow allow users to opt in to the no-inline version (and suggest that they do if we suspect the whole loop gets optimized away), or would that muddy the clear semantic reasoning we are building here?

**Comment by mbauman** [comment]
~~I can go either way here~~ (actually, I think I just convinced myself), but we definitely should not do or even allow both.

If we use a noinline function, there is simply always a constant function call overhead.  This is then comparable.  E.g., if I see 3ns, I know there was one function call + minimal work (the function I was testing got inlined into the noinline test function).  6ns is two function calls and minimal work (it didn't inline!).

Without it, everything is one function-call-overhead faster.  So you can still compare an inlined result with a function that didn't inline.  Well, sort-of.  Inlined functions could behave in absolutely _any_ way.  LLVM can do all sorts of wonky optimizations, ranging from full loop elision to hoisting of some work to doing nothing at all.  And this is an interaction with the testing loop, not with the function call itself.

So let's reinstate the `@noinline` barrier.  Sound good?

**Comment by staticfloat** [comment]
Yes, I agree. It's better to have reliable test machinery than fast test
machinery.

**Comment by sglyon** [comment]
also agree

**Comment by mbauman** [comment]
Alright, updated to use inner functions.  I've also used gensyms for the names just to be safe since I'm not 100% sure how hygiene is working with nested macros calling inline macros.

File: src/benchmarkable.jl [hunk 2]
```diff
@@ -27,33 +27,39 @@
...
-    # in order to ensure we aren't timing untyped dispatch lookup
-    nonconst_locals = map(gensym, nonconst)
-    decls = Expr(:block, map((r,s)->:(const $r = $(esc(s))), nonconst_locals, nonconst)...)
+    # expand the passed expression to support `x[1]` and macros
+    expr = expand(core)
+    # only support function calls
+    expr.head == :call || throw(ArgumentError("expression to benchmark must be a function call"))
+    f = expr.args[1]
+    nargs = length(expr.args) - 1
+    args = Symbol[gensym("arg_$i") for i=1:nargs]
```
Comment (johnmyleswhite) [inline src/benchmarkable.jl]: My heart breaks every time I see an `i=1:nargs` when we have a second idiom that does the exact same thing. :(

Comment (mbauman) [inline src/benchmarkable.jl]: I will forever be tainted by my dirty history with Matlab.

Comment (johnmyleswhite) [inline src/benchmarkable.jl]: I think my dislike for this idiom is mostly attributable to me associating it with Matlab.

**Comment by staticfloat** [comment]
I tested this out, and it seems to work for my benchmarks.  Thanks Matt!

**Comment by johnmyleswhite** [comment]
Ok, I'm pretty happy with this design. Rebase and merge?

**Comment by mbauman** [comment]
Done and done!

**Pull request merged**